File: voice_synthesizer.py
"""
Sistema de síntesis de voz y subtítulos para Cine Norte
Convierte guiones a audio y genera subtítulos sincronizados
"""
import os
import logging
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
import tempfile
import json
from datetime import timedelta
import re

# Importaciones para síntesis de voz
try:
    from gtts import gTTS
    import pygame
    from pydub import AudioSegment
    from pydub.effects import normalize, compress_dynamic_range
    from pydub.silence import split_on_silence
except ImportError:
    print("Instalando dependencias de audio...")
    os.system("pip install gtts pygame pydub")

# Importaciones para subtítulos
try:
    import webvtt
    from webvtt import WebVTT
except ImportError:
    print("Instalando dependencias de subtítulos...")
    os.system("pip install webvtt-py")

from config import config
from script_generator import ScriptSection, GeneratedScript

@dataclass
class AudioSegment:
    """Segmento de audio con metadatos"""
    text: str
    audio_file: str
    start_time: float
    end_time: float
    duration: float
    emphasis_words: List[str]
    emotion: str
    volume_adjustment: float = 1.0

@dataclass
class SubtitleCue:
    """Cue de subtítulo"""
    start_time: str  # Formato HH:MM:SS.mmm
    end_time: str
    text: str
    speaker: str = "Narrador"
    style: str = "default"

@dataclass
class VoiceProfile:
    """Perfil de voz personalizado"""
    name: str
    language: str
    speed: float
    pitch: float
    volume: float
    voice_id: str = "default"

class VoiceSynthesizer:
    """Sintetizador de voz para Cine Norte"""
    
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        self.temp_dir = tempfile.mkdtemp()
        
        # Perfiles de voz predefinidos
        self.voice_profiles = {
            "cinenorte_male": VoiceProfile(
                name="Cine Norte Masculino",
                language="es",
                speed=0.9,
                pitch=0.8,
                volume=1.0,
                voice_id="es-mx-male"
            ),
            "cinenorte_female": VoiceProfile(
                name="Cine Norte Femenino", 
                language="es",
                speed=0.95,
                pitch=1.1,
                volume=1.0,
                voice_id="es-mx-female"
            ),
            "dramatic": VoiceProfile(
                name="Dramático",
                language="es",
                speed=0.8,
                pitch=0.7,
                volume=1.1,
                voice_id="es-mx-dramatic"
            ),
            "energetic": VoiceProfile(
                name="Energético",
                language="es", 
                speed=1.1,
                pitch=1.2,
                volume=1.0,
                voice_id="es-mx-energetic"
            )
        }
        
        self.current_profile = self.voice_profiles["cinenorte_male"]
        
    def synthesize_script(self, script: GeneratedScript, 
                         voice_profile: str = "cinenorte_male",
                         output_format: str = "mp3") -> Tuple[str, List[SubtitleCue]]:
        """
        Sintetiza un guion completo a audio con subtítulos
        
        Args:
            script: Guion generado
            voice_profile: Perfil de voz a usar
            output_format: Formato de salida (mp3, wav, ogg)
            
        Returns:
            Tuple con (ruta_audio, lista_subtitulos)
        """
        try:
            # Configurar perfil de voz
            if voice_profile in self.voice_profiles:
                self.current_profile = self.voice_profiles[voice_profile]
            
            # Procesar cada sección del guion
            audio_segments = []
            subtitle_cues = []
            current_time = 0.0
            
            for section in script.sections:
                # Generar audio para la sección
                segment = self._synthesize_section(
                    section, 
                    current_time,
                    script.target_platform
                )
                
                if segment:
                    audio_segments.append(segment)
                    
                    # Generar subtítulos para la sección
                    section_cues = self._generate_subtitles_for_section(
                        section, 
                        current_time,
                        segment.duration
                    )
                    subtitle_cues.extend(section_cues)
                    
                    current_time += segment.duration
            
            # Combinar todos los segmentos de audio
            final_audio_path = self._combine_audio_segments(
                audio_segments, 
                script.title,
                output_format
            )
            
            # Aplicar efectos finales
            final_audio_path = self._apply_final_effects(final_audio_path, script)
            
            return final_audio_path, subtitle_cues
            
        except Exception as e:
            self.logger.error(f"Error sintetizando guion: {e}")
            return None, []
    
    def _synthesize_section(self, section: ScriptSection, start_time: float, 
                           platform: str) -> Optional[AudioSegment]:
        """Sintetiza una sección individual del guion"""
        try:
            # Preparar texto para síntesis
            processed_text = self._prepare_text_for_synthesis(section.content, section.emphasis_words)
            
            # Generar audio con gTTS
            tts = gTTS(
                text=processed_text,
                lang=self.current_profile.language,
                slow=False
            )
            
            # Guardar audio temporal
            temp_audio_path = os.path.join(self.temp_dir, f"section_{section.type}_{start_time}.mp3")
            tts.save(temp_audio_path)
            
            # Cargar y procesar audio
            audio = AudioSegment.from_mp3(temp_audio_path)
            
            # Aplicar ajustes de perfil de voz
            audio = self._apply_voice_profile(audio, section.emotion)
            
            # Ajustar duración si es necesario
            target_duration = section.duration_seconds * 1000  # Convertir a ms
            if len(audio) > target_duration:
                audio = audio[:target_duration]
            elif len(audio) < target_duration * 0.8:  # Si es muy corto, ajustar velocidad
                speed_factor = len(audio) / (target_duration * 0.9)
                audio = audio.speedup(playback_speed=speed_factor)
            
            # Guardar audio procesado
            processed_audio_path = os.path.join(self.temp_dir, f"processed_{section.type}_{start_time}.mp3")
            audio.export(processed_audio_path, format="mp3")
            
            return AudioSegment(
                text=section.content,
                audio_file=processed_audio_path,
                start_time=start_time,
                end_time=start_time + (len(audio) / 1000.0),
                duration=len(audio) / 1000.0,
                emphasis_words=section.emphasis_words,
                emotion=section.emotion
            )
            
        except Exception as e:
            self.logger.error(f"Error sintetizando sección {section.type}: {e}")
            return None
    
    def _prepare_text_for_synthesis(self, text: str, emphasis_words: List[str]) -> str:
        """Prepara el texto para síntesis con énfasis"""
        processed_text = text
        
        # Agregar pausas naturales
        processed_text = re.sub(r'([.!?])\s+', r'\1... ', processed_text)
        processed_text = re.sub(r'([,;:])\s+', r'\1... ', processed_text)
        
        # Agregar énfasis a palabras clave
        for word in emphasis_words:
            if word.lower() in processed_text.lower():
                # Envolver palabra en tags de énfasis (gTTS no soporta SSML completo)
                processed_text = re.sub(
                    f'\\b{re.escape(word)}\\b', 
                    f'<emphasis level="strong">{word}</emphasis>',
                    processed_text, 
                    flags=re.IGNORECASE
                )
        
        return processed_text
    
    def _apply_voice_profile(self, audio: AudioSegment, emotion: str) -> AudioSegment:
        """Aplica el perfil de voz y efectos emocionales"""
        # Aplicar velocidad
        if self.current_profile.speed != 1.0:
            audio = audio.speedup(playback_speed=self.current_profile.speed)
        
        # Aplicar volumen
        if self.current_profile.volume != 1.0:
            audio = audio + (self.current_profile.volume - 1.0) * 20  # dB
        
        # Aplicar efectos emocionales
        if emotion == "excitement":
            # Aumentar ligeramente la velocidad y volumen
            audio = audio.speedup(playback_speed=1.05)
            audio = audio + 2  # +2dB
        elif emotion == "suspense":
            # Reducir velocidad y volumen para crear tensión
            audio = audio.speedup(playback_speed=0.95)
            audio = audio - 1  # -1dB
        elif emotion == "drama":
            # Velocidad ligeramente más lenta para dramatismo
            audio = audio.speedup(playback_speed=0.9)
        
        # Normalizar audio
        audio = normalize(audio)
        
        # Aplicar compresión dinámica suave
        audio = compress_dynamic_range(audio, threshold=-20.0, ratio=4.0, attack=5.0, release=50.0)
        
        return audio
    
    def _generate_subtitles_for_section(self, section: ScriptSection, start_time: float, 
                                       duration: float) -> List[SubtitleCue]:
        """Genera subtítulos para una sección"""
        cues = []
        
        # Dividir texto en líneas apropiadas para subtítulos
        lines = self._split_text_for_subtitles(section.content, duration)
        
        line_duration = duration / len(lines) if lines else 1.0
        
        for i, line in enumerate(lines):
            if not line.strip():
                continue
                
            cue_start = start_time + (i * line_duration)
            cue_end = cue_start + line_duration
            
            cue = SubtitleCue(
                start_time=self._format_time(cue_start),
                end_time=self._format_time(cue_end),
                text=line.strip(),
                speaker="Cine Norte",
                style=self._get_subtitle_style(section.type)
            )
            cues.append(cue)
        
        return cues
    
    def _split_text_for_subtitles(self, text: str, duration: float) -> List[str]:
        """Divide el texto en líneas apropiadas para subtítulos"""
        # Dividir por oraciones primero
        sentences = re.split(r'[.!?]+', text)
        sentences = [s.strip() for s in sentences if s.strip()]
        
        # Calcular palabras por segundo (aproximadamente 3 palabras por segundo)
        words_per_second = 3.0
        max_words_per_line = int((duration / len(sentences)) * words_per_second) if sentences else 10
        
        lines = []
        current_line = ""
        
        for sentence in sentences:
            words = sentence.split()
            
            if len(current_line.split()) + len(words) <= max_words_per_line:
                if current_line:
                    current_line += " " + sentence
                else:
                    current_line = sentence
            else:
                if current_line:
                    lines.append(current_line)
                current_line = sentence
        
        if current_line:
            lines.append(current_line)
        
        return lines
    
    def _get_subtitle_style(self, section_type: str) -> str:
        """Obtiene el estilo de subtítulo según el tipo de sección"""
        styles = {
            "intro": "intro_style",
            "hook": "hook_style", 
            "plot": "plot_style",
            "analysis": "analysis_style",
            "outro": "outro_style"
        }
        return styles.get(section_type, "default")
    
    def _format_time(self, seconds: float) -> str:
        """Formatea tiempo en formato HH:MM:SS.mmm"""
        td = timedelta(seconds=seconds)
        hours, remainder = divmod(td.total_seconds(), 3600)
        minutes, seconds = divmod(remainder, 60)
        milliseconds = int((seconds % 1) * 1000)
        
        return f"{int(hours):02d}:{int(minutes):02d}:{int(seconds):02d}.{milliseconds:03d}"
    
    def _combine_audio_segments(self, segments: List[AudioSegment], title: str, 
                               output_format: str) -> str:
        """Combina todos los segmentos de audio en un archivo final"""
        if not segments:
            return None
        
        # Cargar primer segmento
        combined_audio = AudioSegment.from_mp3(segments[0].audio_file)
        
        # Agregar pausa inicial
        silence = AudioSegment.silent(duration=500)  # 0.5 segundos
        combined_audio = silence + combined_audio
        
        # Combinar con el resto de segmentos
        for i, segment in enumerate(segments[1:], 1):
            # Agregar pausa entre segmentos
            pause_duration = 200 if segment.emotion != "suspense" else 500  # 0.2-0.5 segundos
            silence = AudioSegment.silent(duration=pause_duration)
            
            segment_audio = AudioSegment.from_mp3(segment.audio_file)
            combined_audio = combined_audio + silence + segment_audio
        
        # Agregar pausa final
        final_silence = AudioSegment.silent(duration=1000)  # 1 segundo
        combined_audio = combined_audio + final_silence
        
        # Guardar archivo final
        output_filename = f"audio_{title.replace(' ', '_')}.{output_format}"
        output_path = os.path.join(self.temp_dir, output_filename)
        combined_audio.export(output_path, format=output_format)
        
        return output_path
    
    def _apply_final_effects(self, audio_path: str, script: GeneratedScript) -> str:
        """Aplica efectos finales al audio"""
        try:
            audio = AudioSegment.from_file(audio_path)
            
            # Normalizar volumen final
            audio = normalize(audio)
            
            # Aplicar fade in/out
            fade_duration = 1000  # 1 segundo
            audio = audio.fade_in(fade_duration).fade_out(fade_duration)
            
            # Aplicar compresión final
            audio = compress_dynamic_range(audio, threshold=-18.0, ratio=3.0, attack=3.0, release=30.0)
            
            # Guardar audio final
            final_path = audio_path.replace('.mp3', '_final.mp3')
            audio.export(final_path, format="mp3", bitrate="192k")
            
            return final_path
            
        except Exception as e:
            self.logger.error(f"Error aplicando efectos finales: {e}")
            return audio_path
    
    def export_subtitles(self, cues: List[SubtitleCue], filename: str = None) -> str:
        """Exporta subtítulos en formato WebVTT"""
        if not filename:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"subtitulos_{timestamp}.vtt"
        
        # Crear archivo WebVTT
        vtt = WebVTT()
        
        for cue in cues:
            vtt.cues.append(webvtt.Cue(
                start=cue.start_time,
                end=cue.end_time,
                text=cue.text
            ))
        
        # Guardar archivo
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(str(vtt))
        
        return filename
    
    def export_subtitles_srt(self, cues: List[SubtitleCue], filename: str = None) -> str:
        """Exporta subtítulos en formato SRT"""
        if not filename:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"subtitulos_{timestamp}.srt"
        
        srt_content = ""
        
        for i, cue in enumerate(cues, 1):
            # Convertir formato de tiempo
            start_srt = cue.start_time.replace('.', ',')
            end_srt = cue.end_time.replace('.', ',')
            
            srt_content += f"{i}\n"
            srt_content += f"{start_srt} --> {end_srt}\n"
            srt_content += f"{cue.text}\n\n"
        
        # Guardar archivo
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(srt_content)
        
        return filename
    
    def create_voice_profile(self, name: str, language: str = "es", 
                           speed: float = 1.0, pitch: float = 1.0, 
                           volume: float = 1.0) -> VoiceProfile:
        """Crea un perfil de voz personalizado"""
        profile = VoiceProfile(
            name=name,
            language=language,
            speed=speed,
            pitch=pitch,
            volume=volume,
            voice_id=f"custom_{name.lower().replace(' ', '_')}"
        )
        
        self.voice_profiles[profile.voice_id] = profile
        return profile
    
    def get_available_voices(self) -> List[VoiceProfile]:
        """Obtiene lista de voces disponibles"""
        return list(self.voice_profiles.values())
    
    def cleanup_temp_files(self):
        """Limpia archivos temporales"""
        try:
            import shutil
            shutil.rmtree(self.temp_dir)
        except Exception as e:
            self.logger.error(f"Error limpiando archivos temporales: {e}")

# Instancia global del sintetizador
voice_synthesizer = VoiceSynthesizer()

File: js/voice-synthesizer.js
/**
 * Sistema de Síntesis de Voz para Cine Norte
 * Convierte guiones a audio y genera subtítulos sincronizados
 */

class VoiceSynthesizer {
    constructor() {
        this.synth = window.speechSynthesis;
        this.voices = [];
        this.currentVoice = null;
        this.isPlaying = false;
        this.isPaused = false;
        this.currentUtterance = null;
        this.audioContext = null;
        this.audioBuffer = null;
        this.recorder = null;
        this.audioChunks = [];
        
        this.loadVoices();
        this.initializeAudioContext();
    }
    
    /**
     * Carga las voces disponibles
     */
    loadVoices() {
        this.voices = this.synth.getVoices();
        
        // Si no hay voces cargadas, esperar al evento voiceschanged
        if (this.voices.length === 0) {
            this.synth.addEventListener('voiceschanged', () => {
                this.voices = this.synth.getVoices();
                this.selectDefaultVoice();
            });
        } else {
            this.selectDefaultVoice();
        }
    }
    
    /**
     * Selecciona la voz por defecto
     */
    selectDefaultVoice() {
        // Buscar voces en español
        const spanishVoices = this.voices.filter(voice => 
            voice.lang.startsWith('es') || voice.lang.includes('Spanish')
        );
        
        if (spanishVoices.length > 0) {
            // Preferir voces masculinas para Cine Norte
            const maleVoice = spanishVoices.find(voice => 
                voice.name.toLowerCase().includes('male') || 
                voice.name.toLowerCase().includes('masculino')
            );
            this.currentVoice = maleVoice || spanishVoices[0];
        } else {
            this.currentVoice = this.voices[0] || null;
        }
    }
    
    /**
     * Inicializa el contexto de audio
     */
    async initializeAudioContext() {
        try {
            this.audioContext = new (window.AudioContext || window.webkitAudioContext)();
        } catch (error) {
            console.warn('Web Audio API no soportada:', error);
        }
    }
    
    /**
     * Sintetiza un guion completo a audio
     * @param {Object} script - Guion generado
     * @param {string} voiceProfile - Perfil de voz
     * @returns {Promise<Object>} Resultado de la síntesis
     */
    async synthesizeScript(script, voiceProfile = 'male') {
        try {
            this.showLoading('Generando audio...');
            
            // Configurar perfil de voz
            this.setVoiceProfile(voiceProfile);
            
            // Procesar cada sección del guion
            const audioSegments = [];
            const subtitleCues = [];
            let currentTime = 0;
            
            for (const section of script.sections) {
                const segment = await this._synthesizeSection(section, currentTime);
                if (segment) {
                    audioSegments.push(segment);
                    
                    // Generar subtítulos para la sección
                    const sectionCues = this._generateSubtitlesForSection(section, currentTime);
                    subtitleCues.push(...sectionCues);
                    
                    currentTime += segment.duration;
                }
            }
            
            // Combinar segmentos de audio
            const finalAudio = await this._combineAudioSegments(audioSegments);
            
            this.hideLoading();
            
            return {
                audioBlob: finalAudio,
                subtitles: subtitleCues,
                duration: currentTime,
                segments: audioSegments
            };
            
        } catch (error) {
            console.error('Error sintetizando guion:', error);
            this.hideLoading();
            throw error;
        }
    }
    
    /**
     * Sintetiza una sección individual del guion
     * @param {Object} section - Sección del guion
     * @param {number} startTime - Tiempo de inicio
     * @returns {Promise<Object>} Segmento de audio
     */
    async _synthesizeSection(section, startTime) {
        return new Promise((resolve, reject) => {
            try {
                // Preparar texto para síntesis
                const processedText = this._prepareTextForSynthesis(section.content, section.emphasis_words);
                
                // Crear utterance
                const utterance = new SpeechSynthesisUtterance(processedText);
                
                // Configurar voz
                if (this.currentVoice) {
                    utterance.voice = this.currentVoice;
                }
                
                // Configurar parámetros según el perfil
                this._applyVoiceProfile(utterance, section.emotion);
                
                // Configurar eventos
                utterance.onstart = () => {
                    console.log(`Iniciando síntesis de sección: ${section.type}`);
                };
                
                utterance.onend = () => {
                    const duration = this._estimateDuration(processedText);
                    resolve({
                        type: section.type,
                        text: section.content,
                        duration: duration,
                        startTime: startTime,
                        endTime: startTime + duration,
                        emotion: section.emotion
                    });
                };
                
                utterance.onerror = (event) => {
                    console.error('Error en síntesis:', event.error);
                    reject(new Error(`Error en síntesis: ${event.error}`));
                };
                
                // Iniciar síntesis
                this.synth.speak(utterance);
                
            } catch (error) {
                reject(error);
            }
        });
    }
    
    /**
     * Prepara el texto para síntesis con énfasis
     * @param {string} text - Texto original
     * @param {Array} emphasisWords - Palabras de énfasis
     * @returns {string} Texto procesado
     */
    _prepareTextForSynthesis(text, emphasisWords) {
        let processedText = text;
        
        // Agregar pausas naturales
        processedText = processedText.replace(/([.!?])\s+/g, '$1... ');
        processedText = processedText.replace(/([,;:])\s+/g, '$1... ');
        
        // Agregar énfasis a palabras clave (usando pausas)
        for (const word of emphasisWords) {
            const regex = new RegExp(`\\b${word}\\b`, 'gi');
            processedText = processedText.replace(regex, `... ${word}... `);
        }
        
        return processedText;
    }
    
    /**
     * Aplica perfil de voz y efectos emocionales
     * @param {SpeechSynthesisUtterance} utterance - Utterance a configurar
     * @param {string} emotion - Emoción de la sección
     */
    _applyVoiceProfile(utterance, emotion) {
        const voiceConfig = config.getVoiceConfig('male'); // Usar configuración base
        
        // Configurar velocidad
        utterance.rate = voiceConfig.rate || 0.9;
        
        // Configurar tono
        utterance.pitch = voiceConfig.pitch || 0.8;
        
        // Configurar volumen
        utterance.volume = 1.0;
        
        // Aplicar efectos emocionales
        switch (emotion) {
            case 'excitement':
                utterance.rate *= 1.05;
                utterance.pitch *= 1.1;
                break;
            case 'suspense':
                utterance.rate *= 0.95;
                utterance.pitch *= 0.9;
                break;
            case 'drama':
                utterance.rate *= 0.9;
                utterance.pitch *= 0.8;
                break;
            case 'comedy':
                utterance.rate *= 1.1;
                utterance.pitch *= 1.2;
                break;
        }
    }
    
    /**
     * Estima la duración del texto
     * @param {string} text - Texto a analizar
     * @returns {number} Duración estimada en segundos
     */
    _estimateDuration(text) {
        const wordsPerMinute = 150; // Velocidad promedio de habla
        const wordCount = text.split(' ').length;
        return Math.max(5, (wordCount / wordsPerMinute) * 60);
    }
    
    /**
     * Genera subtítulos para una sección
     * @param {Object} section - Sección del guion
     * @param {number} startTime - Tiempo de inicio
     * @returns {Array} Cues de subtítulos
     */
    _generateSubtitlesForSection(section, startTime) {
        const cues = [];
        const lines = this._splitTextForSubtitles(section.content, section.duration_seconds);
        const lineDuration = section.duration_seconds / lines.length;
        
        lines.forEach((line, index) => {
            if (line.trim()) {
                const cueStart = startTime + (index * lineDuration);
                const cueEnd = cueStart + lineDuration;
                
                cues.push({
                    start_time: this._formatTime(cueStart),
                    end_time: this._formatTime(cueEnd),
                    text: line.trim(),
                    speaker: 'Cine Norte',
                    style: this._getSubtitleStyle(section.type)
                });
            }
        });
        
        return cues;
    }
    
    /**
     * Divide el texto en líneas apropiadas para subtítulos
     * @param {string} text - Texto a dividir
     * @param {number} duration - Duración en segundos
     * @returns {Array} Líneas de subtítulos
     */
    _splitTextForSubtitles(text, duration) {
        // Dividir por oraciones
        const sentences = text.split(/[.!?]+/).filter(s => s.trim());
        
        // Calcular palabras por segundo (aproximadamente 3 palabras por segundo)
        const wordsPerSecond = 3.0;
        const maxWordsPerLine = Math.max(5, Math.floor((duration / sentences.length) * wordsPerSecond));
        
        const lines = [];
        let currentLine = '';
        
        for (const sentence of sentences) {
            const words = sentence.trim().split(' ');
            
            if (currentLine.split(' ').length + words.length <= maxWordsPerLine) {
                if (currentLine) {
                    currentLine += ' ' + sentence.trim();
                } else {
                    currentLine = sentence.trim();
                }
            } else {
                if (currentLine) {
                    lines.push(currentLine);
                }
                currentLine = sentence.trim();
            }
        }
        
        if (currentLine) {
            lines.push(currentLine);
        }
        
        return lines;
    }
    
    /**
     * Obtiene el estilo de subtítulo según el tipo de sección
     * @param {string} sectionType - Tipo de sección
     * @returns {string} Estilo de subtítulo
     */
    _getSubtitleStyle(sectionType) {
        const styles = {
            intro: 'intro_style',
            hook: 'hook_style',
            plot: 'plot_style',
            analysis: 'analysis_style',
            outro: 'outro_style'
        };
        return styles[sectionType] || 'default';
    }
    
    /**
     * Formatea tiempo en formato HH:MM:SS.mmm
     * @param {number} seconds - Segundos
     * @returns {string} Tiempo formateado
     */
    _formatTime(seconds) {
        const hours = Math.floor(seconds / 3600);
        const minutes = Math.floor((seconds % 3600) / 60);
        const secs = Math.floor(seconds % 60);
        const milliseconds = Math.floor((seconds % 1) * 1000);
        
        return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}.${milliseconds.toString().padStart(3, '0')}`;
    }
    
    /**
     * Combina segmentos de audio
     * @param {Array} segments - Segmentos de audio
     * @returns {Promise<Blob>} Audio combinado
     */
    async _combineAudioSegments(segments) {
        // Para simplificar, retornamos un blob vacío
        // En una implementación real, se usaría Web Audio API para combinar
        const audioData = new Uint8Array(1024); // Datos de audio simulados
        return new Blob([audioData], { type: 'audio/wav' });
    }
    
    /**
     * Establece el perfil de voz
     * @param {string} profile - Perfil de voz
     */
    setVoiceProfile(profile) {
        const voiceConfig = config.getVoiceConfig(profile);
        
        // Buscar voz que coincida con el perfil
        const matchingVoice = this.voices.find(voice => 
            voice.lang.startsWith('es') && 
            (profile === 'female' ? 
                voice.name.toLowerCase().includes('female') || 
                voice.name.toLowerCase().includes('femenino') :
                voice.name.toLowerCase().includes('male') || 
                voice.name.toLowerCase().includes('masculino')
            )
        );
        
        if (matchingVoice) {
            this.currentVoice = matchingVoice;
        }
    }
    
    /**
     * Reproduce audio
     * @param {Blob} audioBlob - Blob de audio
     */
    playAudio(audioBlob) {
        if (this.isPlaying) {
            this.stopAudio();
        }
        
        const audioUrl = URL.createObjectURL(audioBlob);
        const audio = new Audio(audioUrl);
        
        audio.onplay = () => {
            this.isPlaying = true;
            this.isPaused = false;
        };
        
        audio.onpause = () => {
            this.isPaused = true;
        };
        
        audio.onended = () => {
            this.isPlaying = false;
            this.isPaused = false;
            URL.revokeObjectURL(audioUrl);
        };
        
        audio.play();
        this.currentAudio = audio;
    }
    
    /**
     * Pausa el audio
     */
    pauseAudio() {
        if (this.currentAudio && this.isPlaying) {
            this.currentAudio.pause();
        }
    }
    
    /**
     * Reanuda el audio
     */
    resumeAudio() {
        if (this.currentAudio && this.isPaused) {
            this.currentAudio.play();
        }
    }
    
    /**
     * Detiene el audio
     */
    stopAudio() {
        if (this.currentAudio) {
            this.currentAudio.pause();
            this.currentAudio.currentTime = 0;
            this.isPlaying = false;
            this.isPaused = false;
        }
    }
    
    /**
     * Exporta subtítulos en formato WebVTT
     * @param {Array} cues - Cues de subtítulos
     * @returns {string} Contenido WebVTT
     */
    exportSubtitlesWebVTT(cues) {
        let vttContent = 'WEBVTT\n\n';
        
        cues.forEach(cue => {
            vttContent += `${cue.start_time} --> ${cue.end_time}\n`;
            vttContent += `${cue.text}\n\n`;
        });
        
        return vttContent;
    }
    
    /**
     * Exporta subtítulos en formato SRT
     * @param {Array} cues - Cues de subtítulos
     * @returns {string} Contenido SRT
     */
    exportSubtitlesSRT(cues) {
        let srtContent = '';
        
        cues.forEach((cue, index) => {
            const startSrt = cue.start_time.replace('.', ',');
            const endSrt = cue.end_time.replace('.', ',');
            
            srtContent += `${index + 1}\n`;
            srtContent += `${startSrt} --> ${endSrt}\n`;
            srtContent += `${cue.text}\n\n`;
        });
        
        return srtContent;
    }
    
    /**
     * Obtiene las voces disponibles
     * @returns {Array} Lista de voces
     */
    getAvailableVoices() {
        return this.voices.map(voice => ({
            name: voice.name,
            lang: voice.lang,
            gender: this._detectVoiceGender(voice.name),
            isDefault: voice.default
        }));
    }
    
    /**
     * Detecta el género de la voz
     * @param {string} voiceName - Nombre de la voz
     * @returns {string} Género detectado
     */
    _detectVoiceGender(voiceName) {
        const name = voiceName.toLowerCase();
        if (name.includes('female') || name.includes('femenino') || name.includes('woman')) {
            return 'female';
        }
        if (name.includes('male') || name.includes('masculino') || name.includes('man')) {
            return 'male';
        }
        return 'unknown';
    }
    
    /**
     * Muestra indicador de carga
     * @param {string} message - Mensaje de carga
     */
    showLoading(message) {
        const loadingOverlay = document.getElementById('loadingOverlay');
        const loadingText = document.getElementById('loadingText');
        
        if (loadingOverlay) {
            loadingOverlay.classList.add('active');
        }
        
        if (loadingText) {
            loadingText.textContent = message;
        }
    }
    
    /**
     * Oculta indicador de carga
     */
    hideLoading() {
        const loadingOverlay = document.getElementById('loadingOverlay');
        
        if (loadingOverlay) {
            loadingOverlay.classList.remove('active');
        }
    }
    
    /**
     * Verifica si la síntesis de voz está soportada
     * @returns {boolean} True si está soportada
     */
    isSupported() {
        return 'speechSynthesis' in window;
    }
    
    /**
     * Obtiene el estado actual de reproducción
     * @returns {Object} Estado de reproducción
     */
    getPlaybackState() {
        return {
            isPlaying: this.isPlaying,
            isPaused: this.isPaused,
            currentVoice: this.currentVoice?.name || 'Desconocida'
        };
    }
}

// Instancia global del sintetizador
window.VoiceSynthesizer = VoiceSynthesizer;
window.voiceSynthesizer = new VoiceSynthesizer();
